applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <caf/events/ble_common_event.h>
+
+#if CONFIG_DESKTOP_USB_ENABLE
+#include "usb_event.h"
+#endif
+
+#define MODULE dvfs
+#include <caf/events/module_state_event.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_DVFS_LOG_LEVEL);
+
+#include "ld_dvfs_handler.h"
+
+#define DVFS_RETRY_TIMEOUT		K_MSEC(500)
+/* 0x0d00 is a magic number which codes LLPM connection */
+#define REG_CONN_INTERVAL_LLPM_MASK	0x0d00
+
+enum dvfs_state {
+	DVFS_STATE_INITIALIZING,
+	DVFS_STATE_LLPM_CONNECTED,
+	DVFS_STATE_USB_CONNECTED,
+	DVFS_STATE_COUNT
+};
+
+/* Table containing the frequency settings for each state.
+ * It is tailored for nrf54h20dk and might need to be adjusted for other platforms.
+ */
+enum dvfs_frequency_setting dvfs_freq_table[DVFS_STATE_COUNT] = {
+	DVFS_FREQ_HIGH,		/* DVFS_STATE_INITIALIZING */
+	DVFS_FREQ_MEDLOW,	/* DVFS_STATE_LLPM_CONNECTED */
+	DVFS_FREQ_HIGH		/* DVFS_STATE_USB_CONNECTED */
+	};
+
+/* Binary mask tracking which states are requested */
+static uint8_t dfvs_requests_state;
+
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+
+struct DVFS_retry_work_data {
+	struct k_work_delayable DVFS_retry_work;
+	enum dvfs_frequency_setting target_freq;
+	uint8_t dfvs_retries_cnt;
+} static DVFS_retry_data;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static const char *get_dvfs_state_name(enum dvfs_state state)
+{
+	switch (state) {
+	case DVFS_STATE_INITIALIZING: return "DVFS_STATE_INITIALIZING";
+	case DVFS_STATE_LLPM_CONNECTED: return "DVFS_STATE_LLPM_CONNECTED";
+	case DVFS_STATE_USB_CONNECTED: return "DVFS_STATE_USB_CONNECTED";
+	default: return "Unknown";
+	}
+}
+
+static void reschedule_dfvs_retry_work(enum dvfs_frequency_setting target_freq)
+{
+	if (DVFS_retry_data.dfvs_retries_cnt > 3) {
+		LOG_ERR("DVFS initialization failed.");
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+	DVFS_retry_data.dfvs_retries_cnt++;
+	DVFS_retry_data.target_freq = target_freq;
+	k_work_reschedule(&DVFS_retry_data.DVFS_retry_work, DVFS_RETRY_TIMEOUT);
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	size_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret == -EAGAIN) {
+		LOG_DBG("DVFS not initialized, trying again.");
+		reschedule_dfvs_retry_work(target_freq);
+
+	} else if (ret == -EBUSY) {
```

Comment:
if it is -EBUSY then dvfs is already initialized? 
If yes then the log message at the line L75 would be incorrect i think.